pkg/oidc/client/client.go

```diff
@@ -205,6 +206,25 @@ func (c *client) verifyToken(ctx context.Context, token *oauth2.Token, nonce str
 	if nonce != "" && nonce != verifiedIDToken.Nonce {
 		return nil, fmt.Errorf("nonce did not match (wants %s but got %s)", nonce, verifiedIDToken.Nonce)
 	}
+
+	if c.useAccessToken {
+		accessToken, ok := token.Extra("access_token").(string)
+		if !ok {
+			return nil, fmt.Errorf("access_token is missing in the token response: %#v", accessToken)
+		}
+		_, err := verifier.Verify(ctx, accessToken)
```

Comment:
I think there is no need to verify the token on the client. We're not granting the user any access based on the content of the token. Instead we're just passing the token on to the k8s API, where it has to be verified.

Reply:
If we stick with verifying the token, the correct audience (client_id) would have to be passed to the verifier in line 201. The `aud` of the access_token is not necessarily the client_id requesting the token.

Reply:
Addition: The OIDC spec does not mandate the format of an access token. So my comment about the audience when verifying the access token is irrelevant.

[See https://openid.net/specs/openid-connect-core-1_0.html#CodeFlowTokenValidation ](https://openid.net/specs/openid-connect-core-1_0.html#ImplicitTokenValidation) describes the verification of the access token using the `at_hash` claim. This of course requires a valid signature of the id token. Which brings us back to the question whether we should do any verification of the tokens at all, given we're the client and not the resource server.

Reply:
> I think there is no need to verify the token on the client. We're not granting the user any access based on the content of the token. Instead we're just passing the token on to the k8s API, where it has to be verified.

I see your point here. Kubelogin is only passing the token to kubernetes, so we don't gain much by validating it. However, we currently validate the `id_token`, so my inclination is to follow that precedent in this PR. If we want to add a CLI flag to disable local token verification, I think that could make sense. But I'd argue that is out of scope of this particular issue / PR.

 > If we stick with verifying the token, the correct audience (client_id) would have to be passed to the verifier in line 201. The aud of the access_token is not necessarily the client_id requesting the token.
 
Interesting. In my case, the `client_id` is the `aud` for both the `id_token` and the `access_token`. Is this a common case worth supporting? Or do we think it is sufficient to (in a separate PR) add a flag that disables local token validation?

> Addition: The OIDC spec does not mandate the format of an access token. So my comment about the audience when verifying the access token is irrelevant.

> [https://openid.net/specs/openid-connect-core-1_0.html#CodeFlowTokenValidation ](https://openid.net/specs/openid-connect-core-1_0.html#ImplicitTokenValidation)describes the verification of the access token using the at_hash claim. This of course requires a valid signature of the id token. Which brings us back to the question whether we should do any verification of the tokens at all, given we're the client and not the resource server.

Correct me if I'm wrong, but that validation only applies to access tokens acquired from the "authorization" endpoint. In my flow at least, we are using the "Authentication Code" flow. First, we make a call to `/authorize` to optian an authorization code, then we make a call to `/token` to obtain the `id_token` and `access_token`. In my case, the `id_token` will only populate the `at_hash` claim if the token is obtained using the "implicit" flow, where the token is returned directly from the `/authorize` call.

Reply:
> Is this a common case worth supporting?

Many OIDC IdP will issue an access token with the same structure as an OIDC ID Token. But as far as my understanding goes, this is incidental and not mandated by the spec. So validating an access token the same way one would validate an ID token might or might not work. The access token could be an opaque string that only has meaning to the resource server.

If the access token is a JWT after the OIDC spec (as with Azure, Auth0, etc.), the only reasonable value for `aud`, in my opinion, would be the client ID of the resource server because that's who the access token is meant for.

So I would vote for either a) not verify the access token at all or b) only verify it using the `at_hash`, if present. Tendency towards a) because b) implies verifying the ID token, which, in my opinion, is pointless on the client side.

> In my flow at least, we are using the "Authentication Code" flow

Section 3.1.3.8. of the same Document describes the same procedure for the auth code flow. The `at_hash` claim is optional in that case.

Reply:
> So I would vote for either a) not verify the access token at all or b) only verify it using the at_hash, if present. Tendency towards a) because b) implies verifying the ID token, which, in my opinion, is pointless on the client side.

Given that we currently verify the ID token in this application and changing that is out of scope of this PR, I don't think option a) fits well in here. As I mentioned earlier, I'd like this change to match the existing precedence set by this application, at least until the maintainer indicates otherwise. Perhaps a good in-between we can consider is to check the signature of the access_token but not verify the `aud` claim matches the `client_id`. I think we can accomplish this by creating a different `verifier` for the `access_token`. Here is the one we create for the `id_token`:

https://github.com/int128/kubelogin/blob/f9c3c8aee74ff81784c5d5e1b8c2cbcc76aff8a5/pkg/oidc/client/client.go#L201C25-L201C33

We can instead set `ClientID` to a blank string and set `SkipClientIDCheck` to true. See config parameters [here](https://github.com/coreos/go-oidc/blob/6b94bec0d3f54b3546c663f6c1f4dd2620bf5d49/oidc/verify.go#L79). As the docs for `go-oidc` state, there are some cases where the `aud` will not match the `client_id`. This use of an `access_token` seems to fit that use case

> Section 3.1.3.8. of the same Document describes the same procedure for the auth code flow. The at_hash claim is optional in that case.

Agreed. Using the `at_hash` route for validation is a possibility. But that field is optional, so it won't work to validate the token in all cases. If I understand correctly, this validation route (using `at_hash`) is primarily useful when the `access_token` is **NOT** a signed JWT. If the `access_token` is a signed JWT (which it will have to be to successfully use this code path for kubernetes), then it is preferable to me to validate the JWT the way we are currently (with the potential caveat above about intentionally not checking the `aud` claim).

Reply:
Sounds good to me.

> which it will have to be to successfully use this code path for kubernetes

Very good point!